Add documentation to the plugins. (#888)
llm-d/llm-d-inference-scheduler: pkg/epp/framework/plugins/datalayer/source/models/README.md

# Models Data Source

**Type:** `models-data-source` | **Implementation:** [factories.go](factories.go)

The Models Data Source polls inference server pods for model information and passes the response to a paired [`model-server-protocol-models`](../../extractor/models/README.md) extractor.

> The `"dataLayer"` feature gate must appear in the top-level `featureGates:` list — this is an experimental feature gate defined in [`ExperimentalDatalayerFeatureGate`](../../../../../datalayer/factory.go); without it the EPP ignores the `data:` section entirely, and with it `data: sources:` must be present (missing it causes a startup error).

Reviewer: No longer true. Datalayer is enabled by default.

## What it does

> 1. Iterates over every pod in the `InferencePool`.

Reviewer: ```suggestion
1. Iterates over every ready endpoint assocaited with the `InferencePool`.
```

> 2. Issues a `GET scheme://pod-ip:metricsPort/path` request to each pod.

Reviewer: nit: too specific, IMHO. Can be generalized without using specific paramaters (or if you do, please use `<var>` notation, such as `<scheme>://<endpoint-ip>:<port>/<path>`).
Do you also document the default values for parameters?
Use of `metricsPort` seems wrong for fetching model information. Also unclear how it is made available to the source.

> 3. Parses the OpenAI-compatible `/v1/models` response.

Reviewer: Please confirm. I think it has some vllm specific attributes (e.g., base model of LoRA).

> 4. Forwards the parsed response to any extractors wired to this source via `data: sources:`.

Reviewer: nit: the forwarding is actually handled by the datalayer.Runtime and not the source directly.

## Inputs consumed

- Pod list from the `InferencePool` (polled individually on each scheduling cycle).

## Attributes produced

> None directly. The parsed API response is forwarded to the attached extractors, which store it as endpoint attributes.

Reviewer: this is true of all sources. Only extractors produce attributes.
Consider if we need the same README template for all plugins.

Author: We opened https://github.com/llm-d/llm-d-inference-scheduler/issues/917

## Configuration

- `scheme` (string, optional, default: `"http"`): Protocol scheme: `"http"` or `"https"`.
- `path` (string, optional, default: `"/v1/models"`): URL path for the models API endpoint.
- `insecureSkipVerify` (bool, optional, default: `true`): Skip TLS certificate verification.

```yaml
- type: models-data-source
  name: my-models-source
  parameters:
    scheme: "http"
    path: "/v1/models"
    insecureSkipVerify: true
```

The data source expects responses in the OpenAI-compatible format:

```json
{
  "object": "list",
  "data": [
    { "id": "llama-3-8b", "parent": "llama-3" },
    { "id": "mistral-7b", "parent": "mistral" }
  ]
}
```

## Complete Configuration Example

```yaml
apiVersion: inference.networking.x-k8s.io/v1alpha1
kind: EndpointPickerConfig
featureGates:

> - dataLayer

Reviewer: ```suggestion
```

plugins:
- type: models-data-source
  name: vllm-models-source
  parameters:
    scheme: "https"

>     path: "/v1/models"

Reviewer: Q: should we recommend setting default values?
